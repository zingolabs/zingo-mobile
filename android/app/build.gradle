apply plugin: "com.android.application"
apply plugin: "com.facebook.react"
apply plugin: 'kotlin-android'

/**
 * This is the configuration block to customize your React Native Android app.
 * By default you don't need to apply any configuration, just uncomment the lines you need.
 */

react {
    /* Folders */
    //   The root of your project, i.e. where "package.json" lives. Default is '..'
    // root = file("../")
    //   The folder where the react-native NPM package is. Default is ../node_modules/react-native
    // reactNativeDir = file("../node_modules/react-native")
    //   The folder where the react-native Codegen package is. Default is ../node_modules/@react-native/codegen
    // codegenDir = file("../node_modules/@react-native/codegen")
    //   The cli.js file which is the React Native CLI entrypoint. Default is ../node_modules/react-native/cli.js
    // cliFile = file("../node_modules/react-native/cli.js")
    /* Variants */
    //   The list of variants to that are debuggable. For those we're going to
    //   skip the bundling of the JS bundle and the assets. By default is just 'debug'.
    //   If you add flavors like lite, prod, etc. you'll have to list your debuggableVariants.
    // debuggableVariants = ["liteDebug", "prodDebug"]
    /* Bundling */
    //   A list containing the node command and its flags. Default is just 'node'.
    // nodeExecutableAndArgs = ["node"]
    //
    //   The command to run when bundling. By default is 'bundle'
    // bundleCommand = "ram-bundle"
    //
    //   The path to the CLI configuration file. Default is empty.
    // bundleConfig = file(../rn-cli.config.js)
    //
    //   The name of the generated asset file containing your JS bundle
    // bundleAssetName = "MyApplication.android.bundle"
    //
    //   The entry file for bundle generation. Default is 'index.android.js' or 'index.js'
    // entryFile = file("../js/MyApplication.android.js")
    //
    //   A list of extra flags to pass to the 'bundle' commands.
    //   See https://github.com/react-native-community/cli/blob/main/docs/commands.md#bundle
    // extraPackagerArgs = []
    /* Hermes Commands */
    //   The hermes compiler command to run. By default it is 'hermesc'
    // hermesCommand = "$rootDir/my-custom-hermesc/bin/hermesc"
    //
    //   The list of flags to pass to the Hermes compiler. By default is "-O", "-output-source-map"
    // hermesFlags = ["-O", "-output-source-map"]
}

/**
 * Set this to true to Run Proguard on Release builds to minify the Java bytecode.
 */
def enableProguardInReleaseBuilds = false

/**
 * The preferred build flavor of JavaScriptCore (JSC)
 *
 * For example, to use the international variant, you can use:
 * `def jscFlavor = 'org.webkit:android-jsc-intl:+'`
 *
 * The international variant includes ICU i18n library and necessary data
 * allowing to use e.g. `Date.toLocaleString` and `String.localeCompare` that
 * give correct results when using with locales other than en-US.  Note that
 * this variant is about 6MiB larger per architecture than default.
 */
def jscFlavor = 'org.webkit:android-jsc:+'

/**
 * Whether to enable building a separate APK for each ABI.
 *
 * Defaults to false but can be set to true with the project properties flag
 * e.g. ./gradlew assembleDebug -PsplitApk=true
 *
 * Additional option to include a universal APK
 * e.g. ./gradlew assembleRelease -PsplitApk=true -PincludeUniversalApk=true
 */
ext.splitApk = project.properties['splitApk'] ?: false
ext.includeUniversalApk = project.properties['includeUniversalApk'] ?: false

android {
    namespace 'org.ZingoLabs.Zingo'
    ndkVersion rootProject.ext.ndkVersion

    compileSdkVersion rootProject.ext.compileSdkVersion

    compileOptions {
        sourceCompatibility = "18.0.2.1"
        targetCompatibility = "18.0.2.1"
    }

    kotlinOptions {
        jvmTarget = '18'
    }

    defaultConfig {
        applicationId 'org.ZingoLabs.Zingo' // Real
        //applicationId 'com.ZingoMobile' // @Test
        minSdkVersion rootProject.ext.minSdkVersion
        targetSdkVersion rootProject.ext.targetSdkVersion
        versionCode 126 // Real
        //versionCode 117 // @Test
        versionName "zingo-1.2.5" // Real
        missingDimensionStrategy 'react-native-camera', 'general'
        testBuildType System.getProperty('testBuildType', 'debug')
        testInstrumentationRunner 'androidx.test.runner.AndroidJUnitRunner'
    }

    splits {
        abi {
            enable splitApk.toBoolean()
            reset()
            include "armeabi-v7a", "x86", "arm64-v8a", "x86_64"
            universalApk includeUniversalApk.toBoolean()
        }
    }

    signingConfigs {
        debug {
            storeFile file('debug.keystore')
            storePassword 'android'
            keyAlias 'androiddebugkey'
            keyPassword 'android'
        }
    }

    packagingOptions {
        jniLibs {
            pickFirsts += ['lib/armeabi-v7a/libc++_shared.so', 'lib/arm64-v8a/libc++_shared.so', 'lib/x86/libc++_shared.so', 'lib/x86_64/libc++_shared.so']
        }
    }

    buildTypes {
        debug {
            signingConfig signingConfigs.debug
        }
        release {
            // Caution! In production, you need to generate your own keystore file.
            // see https://reactnative.dev/docs/signed-apk-android.
            signingConfig signingConfigs.debug
            minifyEnabled enableProguardInReleaseBuilds
            proguardFiles getDefaultProguardFile("proguard-android.txt"), "proguard-rules.pro"
            proguardFile "${rootProject.projectDir}/../node_modules/detox/android/detox/proguard-rules-app.pro"
        }
    }

    buildFeatures {
        viewBinding = true
    }

    testOptions {
        // Managed devices for gradle integration testing
        managedDevices {
            devices {
                pixel2api29_x86 (com.android.build.api.dsl.ManagedVirtualDevice) {
                    device = "Pixel 2"
                    apiLevel = 29
                    systemImageSource = "aosp"
                    require64Bit = false
                }
                pixel2api30_x86_64 (com.android.build.api.dsl.ManagedVirtualDevice) {
                    device = "Pixel 2"
                    apiLevel = 30
                    systemImageSource = "aosp"
                    require64Bit = true
                }
            }
            groups {
                x86_Archs {
                    targetDevices.add(devices.pixel2api29_x86)
                    targetDevices.add(devices.pixel2api30_x86_64)
                }
            }
        }
    }
    
    // // applicationVariants are e.g. debug, release
    // applicationVariants.all { variant ->
    //     variant.outputs.each { output ->
    //         // For each separate APK per architecture, set a unique version code as described here:
    //         // https://developer.android.com/studio/build/configure-apk-splits.html
    //         // Example: versionCode 1 will generate 1001 for armeabi-v7a, 1002 for x86, etc.
    //         def versionCodes = ["armeabi-v7a": 1, "x86": 2, "arm64-v8a": 3, "x86_64": 4]
    //         def abi = output.getFilter(OutputFile.ABI)
    //         if (abi != null) {  // null for the universal-debug, universal-release variants
    //             output.versionCodeOverride =
    //                     defaultConfig.versionCode * 1000 + versionCodes.get(abi)
    //         }
    //     }
    // }
}


// Map for the version code that gives each ABI a value.
ext.abiCodes = ['armeabi-v7a':1, 'x86':2, 'arm64-v8a': 3, 'x86_64':4]

import com.android.build.OutputFile

// For each APK output variant, override versionCode with a combination of
// ext.abiCodes * 10000 + variant.versionCode. 
// variant.versionCode is equal to defaultConfig.versionCode.
// If you configure product flavors that define their own versionCode, 
// variant.versionCode uses that value instead.
android.applicationVariants.all { variant ->
  variant.outputs.each { output ->
    def baseAbiVersionCode = project.ext.abiCodes.get(output.getFilter(OutputFile.ABI))

    // Because abiCodes.get() returns null for ABIs that are not mapped by ext.abiCodes,
    // the following code doesn't override the version code for universal APKs.
    // However, because you want universal APKs to have the lowest version code,
    // this outcome is desirable.
    if (baseAbiVersionCode != null) {

      // Assigns the new version code to versionCodeOverride, which changes the
      // version code for only the output APK, not for the variant itself. Skipping
      // this step causes Gradle to use the value of variant.versionCode for the APK.
      output.versionCodeOverride = baseAbiVersionCode * 10000 + variant.versionCode
    }
  }
}
            
dependencies {
    // The version of react-native is set by the React Native Gradle Plugin
    implementation("com.facebook.react:react-android")
    androidTestImplementation('com.wix:detox:+')
    implementation 'androidx.appcompat:appcompat:1.1.0'
    implementation fileTree(dir: "libs", include: ["*.jar"])

    implementation "androidx.swiperefreshlayout:swiperefreshlayout:1.0.0"
    implementation 'com.facebook.soloader:soloader:0.10.5+'

    debugImplementation("com.facebook.flipper:flipper:${FLIPPER_VERSION}") {
      exclude group:'com.facebook.fbjni'
    }

    // Detox tests getAttributes() needs this
    debugImplementation('com.google.android.material:material:1.2.1')

    debugImplementation("com.facebook.flipper:flipper-network-plugin:${FLIPPER_VERSION}") {
        exclude group:'com.facebook.flipper'
        exclude group:'com.squareup.okhttp3', module:'okhttp'
    }

    debugImplementation("com.facebook.flipper:flipper-fresco-plugin:${FLIPPER_VERSION}") {
        exclude group:'com.facebook.flipper'
    }

    if (hermesEnabled) {
        implementation("com.facebook.react:hermes-engine:+") {
            exclude group:'com.facebook.fbjni'
        }
    } else {
        implementation jscFlavor
    }
	implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlinVersion"

    def work_version = "2.7.1"

    // (Java only)
    implementation "androidx.work:work-runtime:$work_version"

    // Kotlin + coroutines
    implementation "androidx.work:work-runtime-ktx:$work_version"

    // optional - RxJava2 support
    implementation "androidx.work:work-rxjava2:$work_version"

    // optional - GCMNetworkManager support
    implementation "androidx.work:work-gcm:$work_version"

    // optional - Test helpers
    androidTestImplementation "androidx.work:work-testing:$work_version"

    // optional - Multiprocess support
    implementation "androidx.work:work-multiprocess:$work_version"
    
    // google truth testing framework
    androidTestImplementation "com.google.truth:truth:1.1.3"
    
    // JSON parsing
    implementation "com.fasterxml.jackson.module:jackson-module-kotlin:2.14.+"
    
    // JUnit test runners
    androidTestImplementation 'junit:junit:4.13'
}

// Run this once to be able to run the application with BUCK
// puts all compile dependencies into folder libs for BUCK to use
task copyDownloadableDepsToLibs(type: Copy) {
    from configurations.implementation
    into 'libs'
}

apply from: file("../../node_modules/@react-native-community/cli-platform-android/native_modules.gradle"); applyNativeModulesAppBuildGradle(project)
repositories {
    mavenCentral()
    google()
}
