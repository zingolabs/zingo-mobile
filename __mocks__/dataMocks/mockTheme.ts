import { ThemeType } from '../../app/types';

export const mockTheme: ThemeType = {
  dark: true,
  colors: {
    background: '#011401', //'#010101',
    card: '#011401', //'#401717',
    border: '#ffffff',
    primary: '#18bd18', //'#df4100',
    primaryDisabled: '#5a8c5a', //'rgba(90, 140, 90, 1)',
    secondaryDisabled: '#233623',
    text: '#c3c3c3',
    zingo: '#888888',
    placeholder: '#888888',
    money: '#ffffff',
    syncing: '#ebff5a',
    notification: '',
    sideMenuBackground: '#040C17',
  },
};
