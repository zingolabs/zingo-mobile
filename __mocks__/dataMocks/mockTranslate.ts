import { CurrencyEnum, DownloadMemosEnum, LanguageEnum } from '../../app/AppState';

export const mockTranslate = (p: string) => {
  if (p === 'about.copyright') {
    return String([
      '1 text translated line 1',
      '2 text translated line 2',
      '3 text translated line 3',
      '4 text translated line 4',
      '5 text translated line 5',
    ]);
  } else if (p === 'seed.buttontexts') {
    return `{
        "new": ["new"],
        "change": ["change"],
        "server": ["server"],
        "view": ["view"],
        "restore": ["restore"],
        "backup": ["backup"]
      }`;
  } else if (p === 'settings.memos') {
    return [
      {
        value: DownloadMemosEnum.noneMemos,
        text: 'text none',
      },
      {
        value: DownloadMemosEnum.walletMemos,
        text: 'text wallet',
      },
      {
        value: DownloadMemosEnum.allMemos,
        text: 'text all',
      },
    ];
  }
  if (p === 'settings.currencies') {
    return [
      {
        value: '',
        text: 'text no currency',
      },
      {
        value: CurrencyEnum.USDCurrency,
        text: 'text USD',
      },
    ];
  }
  if (p === 'settings.languages') {
    return [
      {
        value: LanguageEnum.en,
        text: 'text en',
      },
      {
        value: LanguageEnum.es,
        text: 'text es',
      },
      {
        value: LanguageEnum.pt,
        text: 'text pt',
      },
      {
        value: LanguageEnum.ru,
        text: 'text ru',
      },
    ];
  }
  if (p === 'settings.sendalls') {
    return [
      {
        value: true,
        text: 'text true',
      },
      {
        value: false,
        text: 'text false',
      },
    ];
  }
  if (p === 'settings.donations') {
    return [
      {
        value: true,
        text: 'text true',
      },
      {
        value: false,
        text: 'text false',
      },
    ];
  }
  if (p === 'settings.rescanmenus') {
    return [
      {
        value: true,
        text: 'text true',
      },
      {
        value: false,
        text: 'text false',
      },
    ];
  } else if (p === 'settings.recoverywalletinfoondevices') {
    return [
      {
        value: true,
        text: 'text true',
      },
      {
        value: false,
        text: 'text false',
      },
    ];
  } else {
    return 'text translated';
  }
};
