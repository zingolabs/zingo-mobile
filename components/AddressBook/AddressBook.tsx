/* eslint-disable react-native/no-inline-styles */
import React, { useContext, useState, useEffect, useCallback, useMemo, useRef } from 'react';
import {
  View,
  ScrollView,
  SafeAreaView,
  Keyboard,
  Platform,
  NativeScrollEvent,
  NativeSyntheticEvent,
  TouchableOpacity,
} from 'react-native';
import moment from 'moment';
import 'moment/locale/es';
import 'moment/locale/pt';
import 'moment/locale/ru';

import { useTheme, useScrollToTop } from '@react-navigation/native';
import Animated, { Easing, useSharedValue, withTiming } from 'react-native-reanimated';

import {
  AddressBookActionEnum,
  AddressBookFileClass,
  ButtonTypeEnum,
  GlobalConst,
  SendPageStateClass,
} from '../../app/AppState';
import { ThemeType } from '../../app/types';
import FadeText from '../Components/FadeText';
import Button from '../Components/Button';
import AbDetail from './components/AbDetail';
import AbSummaryLine from './components/AbSummaryLine';
import { ContextAppLoaded } from '../../app/context';
import Header from '../Header';
import AddressBookFileImpl from './AddressBookFileImpl';
import RPC from '../../app/rpc';
import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome';
import { faAnglesUp } from '@fortawesome/free-solid-svg-icons';
import Utils from '../../app/utils';

type AddressBookProps = {
  closeModal: () => void;
  setAddressBook: (ab: AddressBookFileClass[]) => void;
  setSendPageState: (s: SendPageStateClass) => void;
};

const AddressBook: React.FunctionComponent<AddressBookProps> = ({ closeModal, setAddressBook, setSendPageState }) => {
  const context = useContext(ContextAppLoaded);
  const { translate, language, addressBook, addressBookCurrentAddress, addressBookOpenPriorModal, server } = context;
  const { colors } = useTheme() as unknown as ThemeType;
  moment.locale(language);

  const [numAb, setNumAb] = useState<number>(50);
  const [loadMoreButton, setLoadMoreButton] = useState<boolean>(false);
  const [addressBookSorted, setAddressBookSorted] = useState<AddressBookFileClass[]>([]);
  const [addressBookProtected, setAddressBookProtected] = useState<AddressBookFileClass[]>([]);

  const [currentItem, setCurrentItem] = useState<number | null>(null);
  const [titleViewHeight, setTitleViewHeight] = useState<number>(0);
  const [action, setAction] = useState<AddressBookActionEnum | null>(null);
  const [isAtTop, setIsAtTop] = useState<boolean>(true);

  const slideAnim = useSharedValue(0);
  const scrollViewRef = useRef<ScrollView>(null);

  useScrollToTop(scrollViewRef);

  const fetchAddressBookSorted = useMemo(async () => {
    // excluding this address from the list
    const zennyTips = await Utils.getZenniesDonationAddress(server.chainName);
    return addressBook
      .filter((ab: AddressBookFileClass) => ab.address !== zennyTips)
      .sort((a, b) => {
        const aLabel = a.label;
        const bLabel = b.label;
        return aLabel.localeCompare(bLabel);
      })
      .slice(0, numAb);
  }, [addressBook, numAb, server.chainName]);

  const fetchAddressBookProtected = useMemo(async () => {
    // only protected address to use internally ZingoLabs.
    const zennyTips = await Utils.getZenniesDonationAddress(server.chainName);
    return addressBook
      .filter((ab: AddressBookFileClass) => ab.address === zennyTips)
      .sort((a, b) => {
        const aLabel = a.label;
        const bLabel = b.label;
        return aLabel.localeCompare(bLabel);
      });
  }, [addressBook, server.chainName]);

  // because this screen is fired from more places than the menu.
  useEffect(() => {
    (async () => await RPC.rpcSetInterruptSyncAfterBatch(GlobalConst.false))();
  }, []);

  useEffect(() => {
    (async () => {
      const abs = await fetchAddressBookSorted;
      const abp = await fetchAddressBookProtected;
      setLoadMoreButton(numAb < (abs.length || 0));
      setAddressBookSorted(abs);
      setAddressBookProtected(abp);
      // find the current address
      if (addressBookCurrentAddress) {
        const index: number = abs.findIndex((i: AddressBookFileClass) => i.address === addressBookCurrentAddress);
        if (index === -1) {
          setAction(AddressBookActionEnum.Add);
        } else {
          setAction(AddressBookActionEnum.Modify);
        }
        setCurrentItem(index);
      }
    })();
  }, [addressBookCurrentAddress, fetchAddressBookProtected, fetchAddressBookSorted, numAb]);

  useEffect(() => {
    const keyboardDidShowListener = Keyboard.addListener('keyboardDidShow', () => {
      slideAnim.value = withTiming(0 - titleViewHeight + 25, { duration: 100, easing: Easing.linear });
    });
    const keyboardDidHideListener = Keyboard.addListener('keyboardDidHide', () => {
      slideAnim.value = withTiming(0, { duration: 100, easing: Easing.linear });
    });

    return () => {
      !!keyboardDidShowListener && keyboardDidShowListener.remove();
      !!keyboardDidHideListener && keyboardDidHideListener.remove();
    };
  }, [slideAnim, titleViewHeight]);

  const loadMoreClicked = useCallback(() => {
    setNumAb(numAb + 50);
  }, [numAb]);

  const newAddressBookItem = () => {
    setCurrentItem(-1);
    setAction(AddressBookActionEnum.Add);
  };

  const cancel = () => {
    setCurrentItem(null);
    setAction(null);
    if (addressBookCurrentAddress) {
      closeModal();
      setTimeout(
        () => {
          addressBookOpenPriorModal();
        },
        Platform.OS === GlobalConst.platformOSios ? 100 : 1,
      );
    }
  };

  const doAction = async (a: AddressBookActionEnum, label: string, address: string) => {
    if (!label || !address) {
      return;
    }
    let ab: AddressBookFileClass[] = [];
    if (a === AddressBookActionEnum.Delete) {
      ab = await AddressBookFileImpl.removeAddressBookItem(label, address);
    } else {
      ab = await AddressBookFileImpl.writeAddressBookItem(label, address);
    }
    setAddressBook(ab);
    cancel();
  };

  const handleScrollToTop = () => {
    if (scrollViewRef.current) {
      scrollViewRef.current.scrollTo({ y: 0, animated: true });
    }
  };

  const handleScroll = (event: NativeSyntheticEvent<NativeScrollEvent>) => {
    const { contentOffset } = event.nativeEvent;
    const isTop = contentOffset.y === 0;
    setIsAtTop(isTop);
  };

  //console.log('render Address Book - 4', currentItem, action);

  return (
    <SafeAreaView
      style={{
        display: 'flex',
        justifyContent: 'flex-start',
        alignItems: 'stretch',
        height: '100%',
        backgroundColor: colors.background,
      }}>
      <Animated.View style={{ marginTop: slideAnim }}>
        <View
          onLayout={e => {
            const { height } = e.nativeEvent.layout;
            setTitleViewHeight(height);
          }}>
          <Header
            title={translate('addressbook.title') as string}
            noBalance={true}
            noSyncingStatus={true}
            noDrawMenu={true}
            noPrivacy={true}
          />
        </View>
      </Animated.View>

      <ScrollView
        ref={scrollViewRef}
        onScroll={handleScroll}
        scrollEventThrottle={100}
        testID="addressbook.scroll-view"
        keyboardShouldPersistTaps="handled"
        style={{ maxHeight: '85%' }}
        contentContainerStyle={{
          flexDirection: 'column',
          alignItems: 'stretch',
          justifyContent: 'flex-start',
        }}>
        {currentItem === -1 && action !== null && (
          <AbDetail
            index={-1}
            key={'detail-new'}
            item={{} as AddressBookFileClass}
            cancel={cancel}
            action={action}
            doAction={doAction}
            addressBookCurrentAddress={addressBookCurrentAddress}
          />
        )}
        {currentItem !== null && currentItem > -1 && action !== null && (
          <AbDetail
            index={currentItem}
            key={`detail-${currentItem}-${addressBookSorted[currentItem].label}`}
            item={addressBookSorted[currentItem]}
            cancel={cancel}
            action={action}
            doAction={doAction}
          />
        )}
        {!addressBookCurrentAddress && addressBookSorted.length === 0 && currentItem !== -1 && (
          <View
            style={{
              height: 150,
              display: 'flex',
              alignItems: 'center',
              justifyContent: 'flex-start',
              marginTop: 30,
            }}>
            <FadeText style={{ color: colors.primary }}>{translate('addressbook.empty') as string}</FadeText>
          </View>
        )}
        {!addressBookCurrentAddress &&
          addressBookSorted.flatMap((aBItem, index) => {
            return (
              <View key={`container-${index}-${aBItem.label}`}>
                {currentItem === index && (
                  <AbSummaryLine
                    index={index}
                    key={`line-${index}-${aBItem.label}`}
                    item={aBItem}
                    setCurrentItem={setCurrentItem}
                    setAction={setAction}
                    setSendPageState={setSendPageState}
                    closeModal={closeModal}
                    handleScrollToTop={handleScrollToTop}
                    doAction={doAction}
                  />
                )}
              </View>
            );
          })}
        {!addressBookCurrentAddress &&
          addressBookSorted.flatMap((aBItem, index) => {
            return (
              <View key={`container-${index}-${aBItem.label}`}>
                {currentItem !== index && (
                  <AbSummaryLine
                    index={index}
                    key={`line-${index}-${aBItem.label}`}
                    item={aBItem}
                    setCurrentItem={setCurrentItem}
                    setAction={setAction}
                    setSendPageState={setSendPageState}
                    closeModal={closeModal}
                    handleScrollToTop={handleScrollToTop}
                    doAction={doAction}
                  />
                )}
              </View>
            );
          })}
        {!addressBookCurrentAddress &&
          addressBookProtected.flatMap((aBItem, index) => {
            return (
              <View key={`container-${index}-${aBItem.label}`}>
                <AbSummaryLine
                  index={index}
                  key={`line-${index}-${aBItem.label}`}
                  item={aBItem}
                  setCurrentItem={setCurrentItem}
                  setAction={setAction}
                  setSendPageState={setSendPageState}
                  closeModal={closeModal}
                  handleScrollToTop={handleScrollToTop}
                  doAction={doAction}
                  addressProtected={true}
                />
              </View>
            );
          })}
        {loadMoreButton ? (
          <View
            style={{
              height: 150,
              display: 'flex',
              alignItems: 'center',
              justifyContent: 'flex-start',
              marginTop: 5,
              marginBottom: 30,
            }}>
            <Button
              type={ButtonTypeEnum.Secondary}
              title={translate('addressbook.loadmore') as string}
              onPress={loadMoreClicked}
            />
          </View>
        ) : (
          <>
            {!addressBookCurrentAddress && !!addressBookSorted && !!addressBookSorted.length && (
              <View
                style={{
                  height: 150,
                  display: 'flex',
                  alignItems: 'center',
                  justifyContent: 'flex-start',
                  marginTop: 5,
                  marginBottom: 30,
                }}>
                <FadeText style={{ color: colors.primary }}>{translate('addressbook.end') as string}</FadeText>
              </View>
            )}
          </>
        )}
      </ScrollView>
      {!isAtTop && (
        <TouchableOpacity onPress={handleScrollToTop} style={{ position: 'absolute', bottom: 70, right: 10 }}>
          <FontAwesomeIcon
            style={{ marginLeft: 5, marginRight: 5, marginTop: 0 }}
            size={50}
            icon={faAnglesUp}
            color={colors.zingo}
          />
        </TouchableOpacity>
      )}
      {currentItem === null && (
        <View
          style={{
            flexGrow: 1,
            flexDirection: 'row',
            justifyContent: 'center',
            alignItems: 'center',
            marginVertical: 5,
          }}>
          <Button
            testID="addressbook.button.new"
            type={ButtonTypeEnum.Primary}
            title={translate('addressbook.new') as string}
            onPress={() => newAddressBookItem()}
          />
          <Button
            type={ButtonTypeEnum.Secondary}
            title={translate('cancel') as string}
            style={{ marginLeft: 10 }}
            onPress={closeModal}
          />
        </View>
      )}
    </SafeAreaView>
  );
};

export default React.memo(AddressBook);
