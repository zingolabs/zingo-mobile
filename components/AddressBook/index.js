import AddressBook from './AddressBook';
import AddressBookFileImpl from './AddressBookFileImpl';

export { AddressBook, AddressBookFileImpl };
