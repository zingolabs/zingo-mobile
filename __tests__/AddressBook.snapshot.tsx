/**
 * @format
 */

import 'react-native';
import React from 'react';

import { render } from '@testing-library/react-native';
import { ContextAppLoadedProvider, defaultAppContextLoaded } from '../app/context';
import { AddressBook } from '../components/AddressBook';
import { mockAddressBook } from '../__mocks__/dataMocks/mockAddressBook';
import { mockTranslate } from '../__mocks__/dataMocks/mockTranslate';
import { mockTheme } from '../__mocks__/dataMocks/mockTheme';

jest.useFakeTimers();
jest.mock('@fortawesome/react-native-fontawesome', () => ({
  FontAwesomeIcon: '',
}));
jest.mock('react-native-localize', () => ({
  getNumberFormatSettings: () => {
    return {
      decimalSeparator: '.', // us
      groupingSeparator: ',', // us
    };
  },
}));
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');
jest.mock('@react-native-community/netinfo', () => {
  const RN = jest.requireActual('react-native');

  RN.NativeModules.RNCNetInfo = {
    execute: jest.fn(() => '{}'),
  };

  return RN;
});
jest.mock('react-native', () => {
  const RN = jest.requireActual('react-native');

  RN.NativeModules.RPCModule = {
    execute: jest.fn(() => '{}'),
  };

  return RN;
});
jest.mock('@react-navigation/native', () => ({
  useScrollToTop: jest.fn(),
  useTheme: () => mockTheme,
}));

// test suite
describe('Component Address Book - test', () => {
  //snapshot test
  test('Address Book - snapshot', () => {
    const state = defaultAppContextLoaded;
    state.addressBook = mockAddressBook;
    state.translate = mockTranslate;
    const onClose = jest.fn();
    const onSet = jest.fn();
    const ab: any = render(
      <ContextAppLoadedProvider value={state}>
        <AddressBook closeModal={onClose} setAddressBook={onSet} setSendPageState={onSet} />
      </ContextAppLoadedProvider>,
    );
    expect(ab.toJSON()).toMatchSnapshot();
  });
});
