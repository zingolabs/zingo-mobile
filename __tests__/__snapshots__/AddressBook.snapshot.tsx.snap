// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Address Book - test Address Book - snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "alignItems": "stretch",
      "backgroundColor": "#011401",
      "display": "flex",
      "height": "100%",
      "justifyContent": "flex-start",
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "alignItems": "stretch",
        "flexDirection": "column",
        "justifyContent": "flex-start",
      }
    }
    keyboardShouldPersistTaps="handled"
    onScroll={[Function]}
    scrollEventThrottle={100}
    style={
      {
        "maxHeight": "85%",
      }
    }
    testID="addressbook.scroll-view"
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "height": 150,
            "justifyContent": "flex-start",
            "marginTop": 30,
          }
        }
      >
        <Text
          style={
            {
              "color": "#18bd18",
              "opacity": 0.65,
            }
          }
        >
          text translated
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "flexGrow": 1,
        "justifyContent": "center",
        "marginVertical": 5,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#18bd18",
          "borderColor": "#c3c3c3",
          "borderRadius": 10,
          "borderWidth": 2,
          "justifyContent": "center",
          "maxWidth": "90%",
          "minHeight": 48,
          "minWidth": "30%",
          "opacity": 1,
          "padding": 0,
          "paddingLeft": 20,
          "paddingRight": 20,
        }
      }
      testID="addressbook.button.new"
    >
      <View
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "center",
            "margin": 0,
            "minHeight": 48,
            "minWidth": 48,
            "padding": 0,
          }
        }
      >
        <Text
          style={
            {
              "color": "#011401",
              "fontSize": 16,
              "fontWeight": "bold",
              "textAlign": "center",
              "textTransform": "uppercase",
            }
          }
        >
          text translated
        </Text>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#011401",
          "borderColor": "#18bd18",
          "borderRadius": 10,
          "borderWidth": 2,
          "justifyContent": "center",
          "marginLeft": 10,
          "maxWidth": "90%",
          "minHeight": 48,
          "minWidth": "30%",
          "opacity": 1,
          "padding": 0,
          "paddingLeft": 20,
          "paddingRight": 20,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "center",
            "margin": 0,
            "minHeight": 48,
            "minWidth": 48,
            "padding": 0,
          }
        }
      >
        <Text
          style={
            {
              "color": "#18bd18",
              "fontSize": 16,
              "fontWeight": "bold",
              "textAlign": "center",
              "textTransform": "uppercase",
            }
          }
        >
          text translated
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
